### ISSUE TITLE
Odroid C1 - v9.7.1 update failed (location error)

### ISSUE THREAD

**Issue by kw-sudo** [MichaIng/DietPi#7200]
Hi,

this morning I got an update advise (_dietpi-update_) for v9.7.1

Did so and got/get this failure message:

#### Details:
- Date           | 2024-08-27 11:02:03
- Program name   | DietPi-Patch
- Command        | `apt-get -y --allow-change-held-packages install linux-image-edge-meson linux-dtb-edge-meson linux-u-boot-odroidc1-edge`
- Exit code      | 100
- DietPi version | v9.7.0 (MichaIng/master)
- Distro version | bookworm (ID=7)
- Kernel version | `Linux DietPi 6.6.36-current-meson #1 SMP Thu Jun 27 11:49:15 UTC 2024 armv7l GNU/Linux`
- Architecture	 | `armhf`
- Hardware model | Odroid C1 (armv7l) (ID=10)
- Power supply   | (5V 2A)
- SD card        | (EG: SanDisk Ultra 64 GB)
#### Steps to reproduce:

1. started `dietpi-update`
2. 
#### Expected behaviour:

- update to v9.7.1
#### Actual behaviour:

- see bug report above
#### Additional logs:
```
Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package linux-u-boot-odroidc1-edge
```

Regards,
Oliver

**Comment by MichaIng** [comment]
Sorry mate, I missed to add the "edge" branch U-Boot package to our APT server. Will take only 5-10 minutes.

**Comment by MichaIng** [comment]
...or less, done. Please retry the update 🙂.

**Comment by MichaIng** [comment]
Hmm, can you show the output of this:
```sh
realpath '/boot/dtb'
ls -l /boot7dtb*
```

**Comment by MichaIng** [comment]
Here the fix: https://github.com/MichaIng/DietPi/commit/371e958

**Comment by MichaIng** [comment]
Thanks for testing, and sorry for not testing that patch well our end. Was a late decision to switch to edge branch, since USB ports do not work on "current" branch, and we were not able to fix this reliably. But we do not own that SBC our end.

**Comment by MichaIng** [comment]
You mean you keep seeing this also after reboot? Then the kernel image seems to not match the kernel modules. While the `dtb` directory was updated/moved correctly, maybe the kernel image was not 🙈. Can you check:
```sh
uname -r
ls -l /boot /lib/modules
```

**Comment by MichaIng** [comment]
My goodness, this is a mess: The kernel "uImage" seems to be edge kernel, but its kernel modules and certain other files are missing, while the "current" kernel files were remaining, which should have been removed along with the `linux-image-current-meson` package removal.

Can you check this:
```sh
dpkg -l | grep linux
dpkg -S /lib/modules/6.* /usr/lib/modules/6.* /boot/config-6.9.7-edge-meson
```

**Comment by MichaIng** [comment]
Found it. Embarrassing mistake our end. Please repeat the update one more time with this command:
```sh
dietpi-update -1
```
This basically re-applies the last update, even that you are on v9.7.1 already.

**Comment by MichaIng** [comment]
Great, now it is as it should be. There is definitely a beta phase needed when we do another kernel migration for Odroid C1, so I can ping some of you known owners of this SBC to test the update, before we release it. Not only do we not have this SBC, but it is also the only one with Armbian-based kernel which still has a `/boot` FAT partition, so a special case in this regards as well.

**Comment by r3n3647** [comment]
Hello there!
I updated  my Odrid C1+ today and now network is not working anymore so I cannot ssh into it anymore.
Can you please give a noob comprehensible instruction how to get it back working?

**Comment by MichaIng** [comment]
Which WiFi adapter do you use? I am asking/wondering, since with the `current` kernel, the USB ports did not work at all, and we only perform the kernel upgrade, if `current` kernel is installed, an neither `edge` or `legacy` concurrently.

Can you show the DietPi update log, just to see whether the kernel update has been applied, and from which previous version on:
```sh
cat /var/tmp/dietpi/logs/dietpi-update.log
```

And WiFi works after you restart the interface?
```sh
systemctl restart ifup@wlan0
```
And can you show the full log for the interface setup attempt after boot?
```sh
journalctl -u ifup@wlan0
```

Maybe it is some timing issue. There are cases where the adapter for whatever reason takes longer to be usable at boot, while the interface is registered already, hence `ifup@wlan0.service` starts and tries to connect, but it cannot work yet. Usually, one would expect the driver to not make the interface available, until the adapter has been fully initialised and is usable, but avoid the need for arbitrary workarounds, but the Odroid C1 is vastly unsupported in mainline kernel and in general, so maybe a recent change in the USB or WiFi driver caused such quirk.

**Comment by r3n3647** [comment]
And WiFi works after you restart the interface?

Yes

And can you show the full log for the interface setup attempt after boot?

[journalctl -u ifup@wlan0](http://0x0.st/Xv-z.txt)


Can you show the DietPi update log, just to see whether the kernel update has been applied, and from which previous version on.

Where is the dietpi-update.log file? 

```
# uname -a
Linux printserver 6.9.7-edge-meson #1 SMP Thu Jun 27 11:52:32 UTC 2024 armv7l GNU/Linux
```

I also have minimal dietpi installation with CUPS only as a printserver.

I did another fresh install on eMMC for odroid c1 and encountered the same issue again - no WLAN after boot. 
Perhaps a race condition.

Meanwhile I had installed and configured a RPi0_I as replacement for my printserver. Works flawless.

**Comment by MichaIng** [comment]
> Where is the dietpi-update.log file?

```sh
cat /var/tmp/dietpi/logs/dietpi-update.log
```

> `wpa_supplicant[437]: nl80211: kernel reports: key addition failed`

Hmm, this error throws some search engine results, but for various different circumstances. Does it help to delay network bring-up a bit?
```sh
mkdir /etc/systemd/system/ifup@wlan0.service.d
echo -e '[Service]\nExecStartPre=/bin/sleep 20' > /etc/systemd/system/ifup@wlan0.service.d/defer.conf
```

> Is the Odroid with this kernel update still promising for a while or should I start looking for a Raspberry?

We were actually about to drop support for Odroid C1 already 5 years ago, when the legacy kernel form Hardkernel did not work well with newly Debian Stretch release, and mainline/Armbian kernel had no GPU/HDMI support. We however kept it as headless system then. Armbian however dropped support soon after, and Hardkernel as well. Some individual volunteers kept fixing things here and there, even HDMI again, but as can be seen from broken USB support with Linux 6.6, it is fragile at best. It runs as long it runs, but any kernel update can break things. I will form now on let every new kernel build be tested by some Odroid C1 users, before distributing it. But if there are more serious regressions, it would mean to be stuck with a non-updated kernel version, which is at some point a security issue as well.

TL;DR: If you have serious tasks, better watch out for a newer SBC.

RPi is always a safe choice, of course, but there are better choices depending on tasks/requirements. Allwinner SoC SBCs are weaker and cheaper, but usually have very quickly and good mainline Linux support, hence long support is moreless guaranteed. Rockchip SoC SBCs can be very powerful and efficient (RK3588), but Rockchip provides an own kernel branch, instead of contributing to mainline kernel, hence it takes longer before proper mainline kernel support is there, with GPU acceleration etc. But they are then usually well maintained as well, since popular.

**Comment by MichaIng** [comment]
I guess the data/configs for those Docker containers can be relatively easily transferred to a new system, if required. I suggest, to keep using the Odroid C1 as long as it works, and is powerful enough to handle the load (no problem with AdGuard Home + Unbound + vaultwarden, you might even consider our slim install options for those 3 software titles). We are doing kernel upgrades and migrations for the Odroid C1 more careful from now on, letting test everything, before shipping it. And you can keep a `dietpi-backup` (or other way of system or SD card backup), just in case some particular hardware has issues after the kernel upgrade, like the WiFi adapter of @r3n3647.

**Comment by r3n3647** [comment]
> Which WiFi adapter do you use?

0bda:8176 Realtek Semiconductor Corp. RTL8188CUS 802.11n WLAN Adapter
Original 'ODROID' branded ;-)

```sh
cat /var/tmp/dietpi/logs/dietpi-update.log
```

File does not exist. (fresh install)

Adding 20s delay to ifup@wlan0.service does not help - connection to WLAN accesspoint is established but still no ip assigned. ifup@wlan0.service already stalls boot for over a minute.

I found a dirty workaround by disabling networking.service, masking ifup@wlan0.service (as something always triggers it also if disabled!) and install / enable dhcpcd with wpa_supplicant hook enabled. It takes quite long till network comes up - but it works for me atm.

The most simple way to 'solve' this issue would be to just downgrade to previous DietPi version 9.6 which was working fine. Unfortunately DietPi upgrades itself to latest version 9.7 after installing as soon as it can reach the internet.

> Is the Odroid with this kernel update still promising for a while or should I start looking for a Raspberry? 

As long as Odroid C1 is supported by any distribution I will try to use it as long as possible. 
Otherwise it would end up as another paperweight like my Odroid U2/U3 devices.
The RaspberryPi ZeroW1 I'am using now as a printserver is very very slow and unstable.

Raspberrys are of course much better maintained than Odroids - but to buy another device as long as I have one which could do the job makes no sense to me.

**Comment by MichaIng** [comment]
@master-kw 
Sorry for the late reply. If `/var/tmp/dietpi/logs/dietpi-update.log` does not exist, if was a DietPi v9.7 image already, right? If you have v9.6 image, could you apply the DietPi update there, and then check above log file? Or maybe easier, just apply an APT upgrade, paste logs (so we see the kernel version change), and check whether WiFi fails as well after reboot:
```sh
apt update
apt upgrade
```

**Comment by MichaIng** [comment]
Oh sorry, you are right 🙂.
